core: Implement RangeVarConstraint

alexarice opened the pull request:
Continuation of #2704 

Co-authored-by: Emilien Bauer <papychacal@gmail.com>

alexarice commented:
Would be nice to get the commit attribution correct as this is still mainly @PapyChacal 's work, I just made pyright happy and put some match statements on the tests

superlopuh commented:
I edited the description, which should propagate authorship when the PR is merged

superlopuh commented on xdsl/irdl/constraints.py:
```diff
@@ -20,6 +20,9 @@ class ConstraintContext:
     """The assignment of constraint variables."""
 
+    ranges: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    """The constraint range variables."""
```
is this also an assignment?

superlopuh replied:
Would be good to have the docstrings be consistent if yes

alexarice replied:
Have changed the docstring

superlopuh reviewed (approved):
Would be great to have @math-fehr and @PapyChacal's ticks before merging

alexarice commented on xdsl/irdl/constraints.py:
```diff
@@ -20,6 +20,9 @@ class ConstraintContext:
+    ranges: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    """The assignment of constraint range variables."""
+
     def copy(self):
```
I've realised that the copy and update method now look incorrect. Am I right in thinking they should also copy/update the ranges? These functions are only used in the `AnyOf` constraint (maybe this constraint can't be used with ranges and so this is fine?

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -20,6 +20,9 @@ class ConstraintContext:
     variables: dict[str, Attribute] = field(default_factory=dict)
     """The assignment of constraint variables."""
 
+    ranges: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
+    """The assignment of constraint range variables."""
```
Should we somehow check somewhere that the variable and range variables have different names?
Otherwise I think the `can_infer` and `infer` funcitons might be incorrect if someone have twice the same name for range variable and variable.

If it's too annoying to check, jsut make an issue for it, as I don't think it's actually gonna happen in practice.

alexarice replied:
I've put this in as the last commit, can revert it if it is deemed too messy

math-fehr commented on xdsl/irdl/constraints.py:
```diff
@@ -20,6 +20,9 @@ class ConstraintContext:
     variables: dict[str, Attribute] = field(default_factory=dict)
     """The assignment of constraint variables."""
 
+    ranges: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
```
```suggestion
    range_variables: dict[str, tuple[Attribute, ...]] = field(default_factory=dict)
```
?

math-fehr reviewed (approved):
Nice!

alexarice commented:
@superlopuh @math-fehr any thoughts on whether to keep the last commit?

superlopuh commented:
IDK, I'd rather add helpers to add and get variables on the original objects, and keep all the logic to one class, from an architecture point of view there's some dubious coupling going on there, it would nice for the unicity logic to be encapsulated.

alexarice commented:
Sure, I'll do that

alexarice commented:
I removed the last commit, implemented it as @superlopuh suggested. In the process of this I rebased past the pyright changes and now had to fix some other stuff. I think this is because I generalised `AnyOf` at some point in this commit mistakenly. I can try to revert these changes if wanted.

math-fehr reviewed (approved):
To me that looks good!
